From af877da17aac047a8c37755879c3dcf127013bc9 Mon Sep 17 00:00:00 2001
From: Caio Lima <ticaiolima@gmail.com>
Date: Mon, 17 Aug 2020 09:01:36 -0300
Subject: [PATCH] [ARMv7][JSC] Conservative GC is not considering `r7` as a
 root https://bugs.webkit.org/show_bug.cgi?id=215512

Reviewed by Yusuke Suzuki.

Since `r7` is a callee-saved register on ARMv7
we need to consider it as a conservative root.

See the statement "A subroutine must preserve
the contents of the registers r4-r8, r10, r11
and SP (and r9 in PCS variants that designate
r9 as v6) form page 15 of
https://developer.arm.com/documentation/ihi0042/f/.

* heap/RegisterState.h:
---
 Source/JavaScriptCore/heap/RegisterState.h | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/Source/JavaScriptCore/heap/RegisterState.h b/Source/JavaScriptCore/heap/RegisterState.h
index b3037dba18c..41de1f689f8 100644
--- a/Source/JavaScriptCore/heap/RegisterState.h
+++ b/Source/JavaScriptCore/heap/RegisterState.h
@@ -78,6 +78,7 @@ struct RegisterState {
     uint32_t r4;
     uint32_t r5;
     uint32_t r6;
+    uint32_t r7;
     uint32_t r8;
     uint32_t r9;
     uint32_t r10;
@@ -92,6 +93,7 @@ struct RegisterState {
     SAVE_REG(r4, registers.r4); \
     SAVE_REG(r5, registers.r5); \
     SAVE_REG(r6, registers.r6); \
+    SAVE_REG(r7, registers.r7); \
     SAVE_REG(r8, registers.r8); \
     SAVE_REG(r9, registers.r9); \
     SAVE_REG(r10, registers.r10); \
-- 
2.20.1

